### ISSUE TITLE
Prevent `{ onClick }` instead of `{ onClick() }`

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#7965]
First of all, sorry for the title. I have no idea how to explain this in short.

```kotlin
@Composable
fun Button(text: String, onClick: () -> Unit) {
  MyButton(Action(text = text) { onClick })
}
```

Do you find the issue? `onClick` is never executed when the lambda inside `Action` is executed. This kind of issues are easy to make and really difficult to spot.

The correct code should be:

```kotlin
@Composable
fun Button(text: String, onClick: () -> Unit) {
  MyButton(Action(text = text) { onClick() })
}
```
or
```kotlin
@Composable
fun Button(text: String, onClick: () -> Unit) {
  MyButton(Action(text = text, onClick = onClick))
}
```

## Expected Behavior of the rule
There are multiple options to spot things like this:

- A lambda that in its signature says that it returns `Unit` but it returns something different.
   - This could create false positives, for example: `{ set.remove("foo") }`. `remove` returns `Boolean` but I think that this code is fine.
-  A lambda that in its signature says that it returns `Unit` but it returns a function/another lambda.
   - Probably better
- Extend `IgnoredReturnValue` and say that the function types can't be ignored. But I think that `IgnoredReturnValue` doesn't work over parameters so I'm not sure if that's a good idea. 

Other case could be:

```kotlin
fun foo(lambda: () -> Unit) {
  lambda
}
```

But if we want to find that we could create a more generic rule because this code has the same problem:
```kotlin
fun foo(bar: String) {
  bar
}
```

So maybe this could end up in two different rules. Right now I'm clueless about their possible names. Any ideas are more than welcome.

## Context
I just reviewed a PR in the project I work on bug fixing an issue like this. I think that detekt can prevent this to happen in the future.

Also this pattern of passing a lambda from one function to another is really common on Compose so I think that this is not and edge case. Also the rule is not Compose specific so we can add it inside our default rule set.

**Comment by CziSKY** [comment]
It appears to be inactive, but with closer inspection, I discover it actually addresses a problem point. (I have made a similar mistake previously 🥹).

Maybe I can take over; here are some implementation ideas I have. Once confirmed, I'll be able to submit a PR within the next 1-2 weeks.

## Case 1

For Case 1, proposing a new rule called `UnitLambdaReturnsFunction`

- Triggered only when the lambda's return type is `Unit` (explicit or inferred).
- Detects when the trailing expression is a "function reference/function type value" and does not have a call (missing `()`).

### Non-compliant example:

```kotlin
// 1) The last statement returns a function type value, not a call
val block: () -> Unit = {
    println("will not run")
    doSomething // Forgot to add (), this just returns a function reference
}

// 2) Misuse in a Jetpack Compose scenario
@Composable
fun MyButton(onClick: () -> Unit) {
    Button(onClick = {
        logEvent // Forgot to add ()
    }) {
        Text("Click")
    }
}

// 3) Nested scenario: The last line is a function reference
fun handle(block: () -> Unit) {
    block()
}

fun demo() {
    handle {
        prepare()
        finalize // Forgot to add ()
    }
}

// 4) Returning a local function as the last statement
val task: () -> Unit = {
    fun runTask() { println("running") }
    runTask // Forgot to add ()
}
```

### Compliant Example:

```kotlin
// 1) Correctly calling the function at the end
val block: () -> Unit = {
    println("will run")
    doSomething()
}

// 2) Jetpack Compose: Ensuring the call happens
@Composable
fun MyButton(onClick: () -> Unit) {
    Button(onClick = {
        logEvent()
    }) {
        Text("Click")
    }
}

// 3) Nested scenario: The function is explicitly called at the end
fun handle(block: () -> Unit) {
    block()
}

fun demo() {
    handle {
        prepare()
        finalize()
    }
}

// 4) The local function is called correctly
val task: () -> Unit = {
    fun runTask() { println("running") }
    runTask()
}
```

### Handling potential false postives:

To address potential false positives (such as `{ set.remove("foo") }`), two new configuration options can be added:

- ignoreInTests (default: true)

Disables the rule in test source sets, where lambda expressions often contain intentional function references or mock setups that shouldn't trigger warnings.

- includeOtherTypes (default: false)

When enabled, extends detection beyond function types to include other potentially useless expressions (like variable references). This may increase false positives but can catch additional issues like `{ someVariable }` where the `variable` value is unused.

Regarding this item, I am not sure whether there should be a whitelist or blacklist to block specific types?

## Case 2

As far as I know, a solution for Case 2 already existed before I posted this comment (e.g., #7967).

After Case 1 is resolved, perhaps I could review the status of Case 2 and open a separate issue to address it?

**Comment by BraisGabin** [comment]
Thanks for your detailed description! It helped my a lot to get my ideas a bit more clear. What I'm thinking is what is the difference between these two?

```kotlin
fun demo() {
    handle {
        prepare()
        finalize // Forgot to add ()
    }
}
```

and 

```kotlin
fun demo() {
    handle {
        finalize // Forgot to add ()
        prepare()
    }
}
```

Both should be reported in the same way, right? So maybe `UnitLambdaReturnsFunction` is just a subset of the issues and we could create something more general. And it looks really similar to what `IgnoredReturnValue` does.

So maybe we should expand `IgnoredReturnValue` to detect things like those. What do you think?

**Comment by CziSKY** [comment]
> Thanks for your detailed description! It helped my a lot to get my ideas a bit more clear. What I'm thinking is what is the difference between these two?
> 
> fun demo() {
>     handle {
>         prepare()
>         finalize // Forgot to add ()
>     }
> }
> and
> 
> fun demo() {
>     handle {
>         finalize // Forgot to add ()
>         prepare()
>     }
> }
> Both should be reported in the same way, right? So maybe `UnitLambdaReturnsFunction` is just a subset of the issues and we could create something more general. And it looks really similar to what `IgnoredReturnValue` does.
> 
> So maybe we should expand `IgnoredReturnValue` to detect things like those. What do you think?

Got it.

And after looking at the code, I also understood the limitations of `IgnoredReturnValue` for `Case 1` - `IgnoredReturnValue` only handles `KtCallExpression` (function calls with parentheses) but not `KtNameReferenceExpression` (e.g. function references without parentheses).

I think there is no problem in extending `IgnoredReturnValue` according to your idea. Can I try it?

**Comment by atulgpt** [comment]
Should we also address issue like in the comment https://github.com/detekt/detekt/pull/8506#discussion_r2290708145
In the commented code I mistakenly added `StandardNames` in one line which was unused(and didn't have any side effect. It was companion class). 
So should we increase the scope of this issue to detekt any unused expression which can't have any side effect?

**Comment by BraisGabin** [comment]
> Can I try it?

Sure! Go for it!

> So should we increase the scope of this issue to detekt any unused expression which can't have any side effect?

That's @CziSKY call. If he wants to implement both in the same pr or just one.

Anyway I agree with you, this rule should catch that too.
